librepo/handle.c

```diff
@@ -841,7 +870,56 @@ lr_handle_setopt(LrHandle *handle,
     return ret;
 }
 
-static gboolean
+gboolean
+lr_handle_network_wait(LrHandle *handle, GError **err, guint seconds)
+{
+
+    assert(!err || *err == NULL);
+
+    if (!handle) {
+        g_set_error(err, LR_HANDLE_ERROR, LRE_BADFUNCARG,
+                    "No handle specified");
+        return FALSE;
+    }
+
+    GNetworkMonitor *monitor = g_network_monitor_get_default();
+    GCancellable *cancellable = g_cancellable_new ();
+
+    struct callback_data data_struct;
+    data_struct.input_seconds = seconds;
+    data_struct.begin = time(NULL);
+    data_struct.cancellable = cancellable;
+    data_struct.monitor = monitor;
+
+    gchar *baseurl;
+    if(handle->metalinkurl)
+      baseurl = handle->metalinkurl;
+    else if(handle->mirrorlisturl)
+      baseurl = handle->mirrorlisturl;
+    else if(handle->urls)
+      baseurl = handle->urls[0];
+    assert(baseurl);
+
+    GUri *uri = g_uri_parse(baseurl, G_URI_FLAGS_NONE, NULL);
+    if(uri == NULL){
+      return FALSE;
+    }
+    const gchar* scheme = g_uri_get_scheme(uri);
+    if(!g_strcmp0(scheme, "file")){
+      return TRUE; 
+    }
+    const gchar* host = g_uri_get_host(uri);
+    guint16 port = g_uri_get_port(uri);
+    GSocketConnectable *connectable =  g_network_address_new(host, port);
+    data_struct.connectable = connectable;
+    g_autoptr(GMainLoop) loop;
+    loop = g_main_loop_new(NULL, FALSE);
+    data_struct.loop = loop;
+    g_signal_connect(monitor, "notify::network-available", G_CALLBACK(on_network_available), &data_struct);    
```

Comment:
I think, what we need to do is query the property first here.  If the network *is* available at the moment then we should proceed directly to waiting on the URL.